### PR TITLE
Support custom claims

### PR DESCRIPTION
**Pull request by litewarp**
Add custom claims to the AuthUser. Also includes the emulator fix (was a pain to debug without it).

Will clean up later, but thought I should push so people can see it.

### CODE REVIEW

File: src/createAuthUser.js [hunk 1]
```diff
@@ -100,20 +102,33 @@ const createAuthUser = ({
...
     tokenString = null
   } else if (firebaseUserAdminSDK) {
+    /**
+     * firebaseUserAdminSDK is a DecodedIDToken obtained from
+     * admin.auth().verifyIdToken which returns all the user's claims
+     * https://firebase.google.com/docs/auth/admin/custom-claims
+     * In order for the claims to be consistent, we need to pass the
+     * entire adminSDK object as claims
+     */
+    claims = { ...firebaseUserAdminSDK }
```
Comment (kmjennison) [inline src/createAuthUser.js:123]: I think we'd want to filter out all standard claims properties here so it includes only custom claims. This will match the client-side implementation and also reduce the size of AuthUser.

File: src/useFirebaseUser.js [hunk 2]
```diff
@@ -3,13 +3,13 @@ import firebase from 'firebase/app'
 import 'firebase/auth'
 import { getConfig } from 'src/config'
 import createAuthUser from 'src/createAuthUser'
+import logDebug from 'src/logDebug'
 
-const defaultTokenChangedHandler = async (authUser) => {
+const defaultTokenChangedHandler = async (authUser, userToken) => {
```
Comment (kmjennison) [inline src/useFirebaseUser.js:8]: Why does this need to also receive an ID token when it already has access via AuthUser?

File: src/createAuthUser.js [hunk 3]
```diff
@@ -83,6 +84,7 @@ const createAuthUser = ({
     )
   }
 
+  let claims = null
```
Comment (kmjennison) [inline src/createAuthUser.js:87]: Default to `{}`.

File: src/useFirebaseUser.js [hunk 4]
```diff
@@ -43,28 +43,53 @@ const defaultTokenChangedHandler = async (authUser) => {
...
+/**
+ * Fetch the currentusers idTokenResult which contains both the idToken and the claims
+ * https://firebase.google.com/docs/reference/js/firebase.auth.IDTokenResult
+ */
+const fetchIdTokenResult = async () => {
+  try {
+    const idTokenResult = await firebase.auth().currentUser.getIdTokenResult()
+    return idTokenResult
+  } catch (e) {
+    logDebug('Error Getting Id Token Result', e)
```
Comment (kmjennison) [inline src/useFirebaseUser.js:55]: I suspect we will just want to throw here instead of swallowing the error.

File: src/useFirebaseUser.js [hunk 5]
```diff
@@ -43,28 +43,53 @@ const defaultTokenChangedHandler = async (authUser) => {
...
+  return defaultTokenChangedHandler(authUser, token)
 }
 
 const useFirebaseUser = () => {
   const [user, setUser] = useState()
   const [initialized, setInitialized] = useState(false)
 
   async function onIdTokenChange(firebaseUser) {
-    setUser(firebaseUser)
+    let idTokenResult = { claims: null, token: null }
```
Comment (kmjennison) [inline src/useFirebaseUser.js:80]: Make sure claims default to an empty object. Either set to `undefined` here (relying on defaults in `createAuthUser`) or `{}`.

File: src/firebaseAdmin.js [hunk 6]
```diff
@@ -5,6 +5,11 @@ import { getConfig } from 'src/config'
 // https://firebase.google.com/docs/auth/admin/errors
 const FIREBASE_ERROR_TOKEN_EXPIRED = 'auth/id-token-expired'
```
Comment (kmjennison) [inline src/firebaseAdmin.js:7]: Please break these into separate PRs.

**Review by kmjennison — CHANGES_REQUESTED** [review]
Thanks for these contributions! This looks great.

To get things cleaned up before merging, please:
* break the emulator and custom claims features into separate PRs
* add full test coverage, including mocking the new Firebase method you use
* check out other comments for my questions and feedback

**Comment by kmjennison** [comment]
Ah, I just saw your other PR for the emulator, #108 👍

File: src/useFirebaseUser.js (hunk 2 above)
Comment (litewarp) [reply src/useFirebaseUser.js:8]: So I can't tell you for sure except that when I removed it and used the get getToken() function on authUser to get the token, I ended up with a number of JWT errors. I'm not sure if this is because the token was undefined (i.e. could authuser be initialized without the right gettoken function?) or if it was because it was malformed. But sending the token that is returned from the getIdTokenResult call works perfectly. Too fried to try and figure it out.

Comment (litewarp) [reply src/useFirebaseUser.js]: Nope, you were right. The reason it was erring for me was because I had a custom tokenhandler in my test app that I forgot to modify. Should be all good now!

File: index.d.ts [hunk 7]
```diff
@@ -9,97 +9,107 @@ import type {
 import type { ComponentType } from 'react'
 import type { ParsedUrlQuery } from 'querystring'
```
Comment (kmjennison) [inline index.d.ts:11]: Please revert these unrelated formatting changes.

File: yarn.lock [hunk 8]
```diff
@@ -1,9887 +0,0 @@
-# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (kmjennison) [inline yarn.lock:1]: Revert to keep this file in source.

File: src/__tests__/createAuthUser.test.js [hunk 9]
```diff
@@ -192,6 +195,31 @@ describe('createAuthUser: firebaseUserAdminSDK', () => {
...
+    })
+  })
+
+  it('includes includes all custom claims that are not standard claims', () => {
+    expect.assertions(1)
+    const createAuthUser = require('src/createAuthUser').default
+    const firebaseUserAdminSDK = {
+      ...createMockFirebaseUserAdminSDK(),
+      foo: 'bar',
+      haz: 'cheese',
```
Comment (kmjennison) [inline src/__tests__/createAuthUser.test.js:208]: 🧀 👍

File: src/testHelpers/authUserInputs.js [hunk 10]
```diff
@@ -22,6 +23,38 @@ export const createMockFirebaseUserAdminSDK = () => ({
   // ... other properties
 })
 
+/**
+ * same as createMockFirebaseUserAdminSDK but you can add claims!
+ * @param {*} claims {Obj} consisting of customClaims
+ * @returns {Obj} a decodedIdToken with extra claim properties
+ */
+export const createMockDecodedIdTokenWithClaims = (claims = {}) => ({
```
Comment (kmjennison) [inline src/testHelpers/authUserInputs.js:31]: Given this is the same as above, it seems unnecessary. Let's remove it, and in the the test that uses it, just spread the result:
```js
const idTokenWithClaims = { ...createMockFirebaseUserAdminSDK(), claims: { stuff: 'here' }}
```

File: src/useFirebaseUser.js [hunk 11]
```diff
@@ -62,9 +64,21 @@ const useFirebaseUser = () => {
...
+    }
+    const firebaseUserWithClaims = {
+      ...firebaseUser,
+      claims: filterStandardClaims(idTokenResult.claims),
+    }
+
+    setUser(firebaseUserWithClaims)
     setInitialized(true)
-    await setAuthCookie(firebaseUser)
+    // pass the idToken to the setAuthCookie function
```
Comment (kmjennison) [inline src/useFirebaseUser.js:80]: Remove outdated comment.

File: src/__tests__/useFirebaseUser.test.js [hunk 12]
```diff
@@ -242,11 +295,12 @@ describe('useFirebaseUser', () => {
...
+    const authUser = {
       ...mockAuthUser,
       getIdToken: expect.any(Function),
       serialize: expect.any(Function),
       signOut: expect.any(Function),
-    })
+    }
+    expect(tokenChangedHandler).toHaveBeenCalledWith(authUser)
   })
 })
```
Comment (kmjennison) [inline src/__tests__/useFirebaseUser.test.js:306]: Please add a test that includes custom claims.

File: src/__tests__/createAuthUser.test.js [hunk 13]
```diff
@@ -125,6 +126,7 @@ describe('createAuthUser: firebaseUserClientSDK', () => {
       firebaseUser: firebaseUserJSSDK,
       signOut: expect.any(Function),
       serialize: expect.any(Function),
+      claims: {},
     })
   })
```
Comment (kmjennison) [inline src/__tests__/createAuthUser.test.js:131]: Please also add tests for the `serializedAuthUser` and `firebaseUserClientSDK` with defined claims properties.

**Review by kmjennison — CHANGES_REQUESTED** [review]
This is looking great, thanks!

**Review by kmjennison — APPROVED** [review]
Well done, and thanks! LGTM.

**Comment by kmjennison** [comment]
Released in [v0.13.0-alpha.2](https://github.com/gladly-team/next-firebase-auth/releases/tag/v0.13.0-alpha.2).
